backend/app/services/pdf_reader.py

```diff
@@ -1,22 +1,52 @@
 from typing import BinaryIO
 from dataclasses import dataclass
 import pymupdf
+import pytesseract
 from io import StringIO
-
-# TODO Marta / Luca: I let you make this better. The signature of the function should be unchanged.
+import re
+from pdf2image import convert_from_bytes
 
 @dataclass
 class PDFReader:
+    file_type: str = "pdf"
     sort: bool = True
+    format_type: str = 'words'
 
     def as_text(self, pdf_data: BinaryIO, start_page: int = 0, end_page: int | None = None) -> str:
-        doc = pymupdf.Document(stream=pdf_data)
+        text = ""
         output = StringIO()
-        pages = [page for page_num, page in enumerate(doc) if page_num>=start_page and (not end_page or page_num<end_page)]
-        for page in pages: # iterate the document pages
-            text = page.get_text(sort=self.sort)
-            output.write(f"{text}\f") # write text of page
-        return output.getvalue()
+        doc = pymupdf.Document(stream=pdf_data)
+        pages = [page_num for page_num, page in enumerate(doc) if page_num>=start_page and (not end_page or page_num<end_page)] 
+        for page_num in pages:
+            page = doc.load_page(page_num)
+            if self.format_type == 'words':
+                page_text = " ".join(elem[4] for elem in page.get_text(self.format_type, sort=self.sort))
+            else: 
+                raise NotImplementedError('format_type not accepted')
+                
+            text += page_text
+
+        if not text.strip():
+            print("No text found with PyMuPDF, using OCR")
+            text = perform_ocr(file_stream=pdf_data)
+        
+        output.write(f"{text}\f") 
+
+        return output.getvalue() 
 
+def perform_ocr(file_stream: BinaryIO) -> str:
+        text = ""
+        try:
```

Comment:
I do not think we want this try except: if there is an error with the text reading we do not want to silence it

Reply:
Yes, you should raise an error here and catch it in the server so that you can return a proper HTTP exception code with a message explaining the error